### PR TITLE
use the built-in etcd proxy to use etcd-mesos if configured

### PR DESCRIPTION
**Pull request by spacejam**
@jdef

### CODE REVIEW

File: images/dcos/bootstrap.sh [hunk 1]
```diff
@@ -163,6 +154,32 @@ EOF
...
 }
 
+prepare_etcd_proxy() {
+  prepare_service ${monitor_dir} ${service_dir} etcd-server ${ETCD_SERVER_RESPAWN_DELAY:-1} << EOF
+#!/bin/sh
+#TODO(jdef) don't run this as root
+#TODO(jdef) would be super-cool to have socket-activation here so that clients can connect before etcd is really ready
+exec 2>&1
+mkdir -p ${etcd_server_data}
+PATH="/opt:${PATH}"
```
Comment (sttts) [inline images/dcos/bootstrap.sh]: Is this path really necessary?

File: images/dcos/bootstrap.sh [hunk 2]
```diff
@@ -71,11 +71,6 @@ controller_manager_host=${CONTROLLER_MANAGER_HOST:-${default_dns_name}}
 controller_manager_port=${CONTROLLER_MANAGER_PORT:-10252}
 echo "* controller manager: $controller_manager_host:$controller_manager_port"
 
-# assume that the leading mesos master is always running a marathon
-# service proxy, perhaps using haproxy.
-service_proxy=${SERVICE_PROXY:-leader.mesos}
```
Comment (sttts) [inline images/dcos/bootstrap.sh]: Why is this removed?

File: images/dcos/bootstrap.sh [hunk 3]
```diff
@@ -84,23 +79,19 @@ echo "* service proxy: $service_proxy"
 etcd_server_port=${ETCD_SERVER_PORT:-4001}
 etcd_server_peer_port=${ETCD_SERVER_PEER_PORT:-4002}
 
-ENABLE_ETCD_SERVER=${ENABLE_ETCD_SERVER:-false}
-if [ "${ENABLE_ETCD_SERVER}" != true ]; then
-  etcd_server_list=${ETCD_SERVER_LIST:-http://${service_proxy}:${etcd_server_port}}
-else
-  etcd_advertise_server_host=${ETCD_ADVERTISE_SERVER_HOST:-127.0.0.1}
-  etcd_server_host=${ETCD_SERVER_HOST:-127.0.0.1}
+ETCD_MESOS=${ETCD_MESOS:-disabled}
```
Comment (sttts) [inline images/dcos/bootstrap.sh]: This variable name is not really intuitive. How about EXTERNAL_ETCD or something in this direction?

File: images/dcos/bootstrap.sh [hunk 4]
```diff
@@ -163,6 +154,32 @@ EOF
   prepare_service_depends etcd-server ${etcd_server_list}/v2/stats/store getsSuccess ${deps}
 }
 
+prepare_etcd_proxy() {
+  prepare_service ${monitor_dir} ${service_dir} etcd-server ${ETCD_SERVER_RESPAWN_DELAY:-1} << EOF
+#!/bin/sh
+#TODO(jdef) don't run this as root
+#TODO(jdef) would be super-cool to have socket-activation here so that clients can connect before etcd is really ready
+exec 2>&1
```
Comment (sttts) [inline images/dcos/bootstrap.sh]: All other service use fdmove instead.

File: images/dcos/bootstrap.sh [hunk 5]
```diff
@@ -163,6 +154,32 @@ EOF
...
+mkdir -p ${etcd_server_data}
```
Comment (sttts) [inline images/dcos/bootstrap.sh]: We don't need this mkdir inside of the service script. Move it outside to simplify the service script.

File: images/dcos/bootstrap.sh [hunk 6]
```diff
@@ -163,6 +154,32 @@ EOF
...
+export PATH
+exec /opt/etcd \\
+  -proxy=on \\
+  -advertise-client-urls ${etcd_advertise_client_urls} \\
+  -data-dir ${etcd_server_data} \\
+  -listen-client-urls ${etcd_listen_client_urls} \\
+  -discovery-srv=${ETCD_MESOS}.mesos \\
+EOF
+
+  local deps="controller-manager scheduler"
```
Comment (sttts) [inline images/dcos/bootstrap.sh]: Does controller-amager depend on etcd?

File: images/dcos/bootstrap.sh (hunk 2 above)
Comment (spacejam) [inline images/dcos/bootstrap.sh]: I could not find a reference to it, and @jdef thought it may be safe to remove.  Do you know otherwise?

Comment (sttts) [inline images/dcos/bootstrap.sh]: You are right. I guess it comes from the marathon service haproxy in an ancient DCOS version.

File: images/dcos/bootstrap.sh (hunk 3 above)
Comment (jdef) [inline images/dcos/bootstrap.sh]: changes in this boostrap file are etcd-mesos specific. the name ETCD_MESOS aligns closely with the dcos config option name that we're adding in multiverse. the goal is to keep the multiverse config options as simple as possible, not to make this bootstrap image as flexible as possible. `disabled` will run an internal etcd server in the uber container. any other value is the name of the etcd-mesos framework that we should attach to.

File: images/dcos/bootstrap.sh (hunk 4 above)
Comment (jdef) [inline images/dcos/bootstrap.sh]: fdmove is part of execlineb. this service script uses `/bin/sh`.

File: images/dcos/bootstrap.sh (hunk 5 above)
Comment (jdef) [inline images/dcos/bootstrap.sh]: disagree. goal should be to keep things the service needs to start in the service script.

File: images/dcos/bootstrap.sh (hunk 1 above)
Comment (jdef) [inline images/dcos/bootstrap.sh]: probably not -- copy/paste from prepare_etcd_server

File: images/dcos/bootstrap.sh (hunk 6 above)
Comment (jdef) [inline images/dcos/bootstrap.sh]: these deps were copied from prepare_etcd_server. at this point the controller manager should not have a direct dependency to etcd so we should be safe to remove it from both this script and prepare_etcd_server.

File: images/dcos/bootstrap.sh (hunk 3 above)
Comment (sttts) [inline images/dcos/bootstrap.sh]: I understand what it does. But the name ETCD_MESOS does not suggest what it does. What is etcd-mesos in the context of a mesos framework?

Comment (jdef) [inline images/dcos/bootstrap.sh]: if `$ETCD_MESOS != "disabled"` then you'll want to set `etcd_server_list` to `http://127.0.0.1:$someport` because it's used (a) as an endpoint for etcd readiness, and (b) by other apps that want to connect to etcd (apiserver, scheduler)

Comment (jdef) [inline images/dcos/bootstrap.sh]: it's a dependency

Comment (jdef) [inline images/dcos/bootstrap.sh]: we can add some short description here outlining the dependency.
unfortunately dependencies in DCOS are currently less than ideal.

File: images/dcos/bootstrap.sh (hunk 4 above)
Comment (sttts) [inline images/dcos/bootstrap.sh]: that's the point. Why are all other scripts using execlineb and this is not?

File: images/dcos/bootstrap.sh (hunk 3 above)
Comment (sttts) [inline images/dcos/bootstrap.sh]: I think you miss my point. I don't argue against the variable, just against the bad name, which is neither a verb nor an adjective, but only two context-less words.

File: images/dcos/bootstrap.sh (hunk 4 above)
Comment (jdef) [inline images/dcos/bootstrap.sh]: using env variables in execlineb sucks. if we get rid of PATH then we could
migrate to execlineb

File: images/dcos/bootstrap.sh (hunk 3 above)
Comment (jdef) [inline images/dcos/bootstrap.sh]: sttts and I settled on ETCD_MESOS_FRAMEWORK_NAME

Comment (spacejam) [inline images/dcos/bootstrap.sh]: sounds good to me :+1:

Comment (spacejam) [inline images/dcos/bootstrap.sh]: `etcd_server_list` is always set, whether the proxy is listening there or the embedded server when ETCD_MESOS_FRAMEWORK_NAME is disabled.  If we want to support external clusters in the future we will want to conditionally set this, but that's out of scope for this change.

File: images/dcos/bootstrap.sh (hunk 4 above)
Comment (sttts) [inline images/dcos/bootstrap.sh]: @spacejam ^^

Comment (spacejam) [inline images/dcos/bootstrap.sh]: This detail is uninteresting to me.

Comment (sttts) [inline images/dcos/bootstrap.sh]: Which means? Nobody knows whether this PATH is needed? Then drop it.

File: images/dcos/bootstrap.sh (hunk 5 above)
Comment (sttts) [inline images/dcos/bootstrap.sh]: It's preparation of the environment. It does not change during the runtime. If it wasn't configurable it would even belong into the Dockerfile. But creating directories for persistent data does not belong into a service script for sure.

**Comment by spacejam** [comment]
@sttts I've nuked the PATH stuff and moved the mkdir outside of the service definition, and switched over to execlineb for consistency.

File: images/dcos/bootstrap.sh [hunk 7]
```diff
@@ -136,25 +127,47 @@ prepare_service_script ${service_dir} .s6-svscan finish <<EOF
...
-export PATH
-exec /opt/etcd \\
-  -advertise-client-urls ${etcd_advertise_client_urls} \\
-  -data-dir ${etcd_server_data} \\
-  -initial-advertise-peer-urls ${etcd_initial_advertise_peer_urls} \\
-  -initial-cluster ${etcd_server_name}=${etcd_initial_advertise_peer_urls} \\
-  -listen-client-urls ${etcd_listen_client_urls} \\
-  -listen-peer-urls ${etcd_listen_peer_urls} \\
+fdmove -c 2 1
+/opt/etcd
```
Comment (sttts) [inline images/dcos/bootstrap.sh]: :+1:

File: images/dcos/bootstrap.sh [hunk 8]
```diff
@@ -71,11 +71,6 @@ controller_manager_host=${CONTROLLER_MANAGER_HOST:-${default_dns_name}}
 controller_manager_port=${CONTROLLER_MANAGER_PORT:-10252}
 echo "* controller manager: $controller_manager_host:$controller_manager_port"
 
-# assume that the leading mesos master is always running a marathon
-# service proxy, perhaps using haproxy.
-service_proxy=${SERVICE_PROXY:-leader.mesos}
-echo "* service proxy: $service_proxy"
```
Comment (sttts) [inline images/dcos/bootstrap.sh]: Can we add similar echo's again printing out which services are created?

**Comment by sttts** [comment]
Only one comment left about echos. Otherwise, lgtm.

**Comment by spacejam** [comment]
ping!

File: images/dcos/bootstrap.sh (hunk 8 above)
Comment (spacejam) [inline images/dcos/bootstrap.sh]: functions.sh:71
```
prepare_service() {         
...                                                                                                                                                                           
  local name="$3"                                                                                                                                                                                                                                                                                                                                     ...                         
  echo "* prepare service ${name}"  
...
}
```

Is this not sufficient?

Comment (sttts) [inline images/dcos/bootstrap.sh]: You are right, didn't remember there was a generic log output.

**Comment by jdef** [comment]
:ship:
